tensorboard/uploader/exporter_test.py

```diff
@@ -227,6 +229,214 @@ def outdir_files():
             },
         )
 
+    def test_e2e_success_case_with_blob_sequence_data(self):
+        """Covers exporting of finished and unfinished blob sequences
+
+        as well as rpc error during blob streaming.
+        """
+        mock_api_client = self._create_mock_api_client()
+
+        def stream_experiments(request, **kwargs):
+            del request  # unused
+            self.assertEqual(kwargs["metadata"], grpc_util.version_metadata())
+
+            response = export_service_pb2.StreamExperimentsResponse()
+            response.experiments.add(experiment_id="123")
+            yield response
+            response = export_service_pb2.StreamExperimentsResponse()
+            response.experiments.add(experiment_id="456")
+            yield response
+
+        def stream_experiment_data(request, **kwargs):
+            self.assertEqual(kwargs["metadata"], grpc_util.version_metadata())
+
+            tag = "__default_graph__"
+            for run in ("train", "test"):
+                response = export_service_pb2.StreamExperimentDataResponse()
+                response.run_name = run
+                response.tag_name = tag
+                display_name = "%s:%s" % (request.experiment_id, tag)
+                response.tag_metadata.CopyFrom(
+                    test_util.scalar_metadata(display_name)
+                )
+                for step in range(1):
+                    response.blob_sequences.steps.append(step)
+                    response.blob_sequences.wall_times.add(
+                        seconds=1571084520 + step, nanos=862939144
+                    )
+                    blob_sequence = blob_pb2.BlobSequence()
+                    if run == "train":
+                        # A finished blob sequence.
+                        blob = blob_pb2.Blob(
+                            blob_id="%s_blob" % run,
+                            state=blob_pb2.BlobState.BLOB_STATE_CURRENT,
+                        )
+                        blob_sequence.entries.append(
+                            blob_pb2.BlobSequenceEntry(blob=blob)
+                        )
+                    elif run == "test":
+                        # An unfinihsed blob.
+                        blob_sequence.entries.append(
+                            # `blob` unspecified: An unfinished blob.
+                            blob_pb2.BlobSequenceEntry()
+                        )
+                    response.blob_sequences.values.append(blob_sequence)
+                yield response
+
+        mock_api_client.StreamExperiments = mock.Mock(wraps=stream_experiments)
+        mock_api_client.StreamExperimentData = mock.Mock(
+            wraps=stream_experiment_data
+        )
+        mock_api_client.StreamBlobData.side_effect = [
```

Comment:
It’s especially hard for me to follow the logic of this stub because
it’s partially defined here and then overwritten with a _different_ mock
later. Can we instead just add the appropriate `test_util.grpc_error`(s)
to this list of side effects, with maybe a brief comment noting that the
first two blobs are for the first experiment and the last two errors for
the second experiment?

Reply:
Acknowledging the readability issue. The difficulty is that I can't find a way to combine return-value side effects and error-raising side effects in the same place. I've added a comment here and above to make it clearer that this will be overridden again below for testing the error condition.

Reply:
Hmm; this test passes for me:

```python
import unittest
from unittest import mock


class Test(unittest.TestCase):
    def test(self):
        m = mock.Mock()
        m.side_effect = [
            "first result",
            ValueError("second exception"),
            iter(("third", "stream", "result")),
        ]
        self.assertEqual(m(), "first result")
        with self.assertRaisesRegex(ValueError, "second exception"):
            m()
        self.assertEqual(list(m()), ["third", "stream", "result"])


if __name__ == "__main__":
    unittest.main()
```

Maybe I’m misunderstanding, but the [docs on `side_effect`][1] suggested
to me that the values of the `side_effect` iterable can be exceptions or
values independently of each other.

[1]: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

Reply:
Thanks for bearing with me. Sorry - there was a superfluous item in the `list` for `side_effect` before. Now that I removed it, it's working. Updated this test. Sorry for the confusion.

Also added an assertion for the call arg. PTAL.